File: DSA/GRAPH/1_Basics/Representation/adjacency_list_graph.cpp
#include <iostream>
#include <vector>
using namespace std;

class Graph {
    int V;
    bool isDirected;
    vector<vector<int>> adj;

public:
    Graph(int V, bool isDirected) {
        this->V = V;
        this->isDirected = isDirected;
        adj.resize(V);
    }

    // Add edge u -> v
    void addEdge(int u, int v) {
        adj[u].push_back(v);
        if (!isDirected)
            adj[v].push_back(u);
    }

    void display() {
        cout << "Adjacency List:\n";
        for (int i = 0; i < V; ++i) {
            cout << i << ": ";
            for (int neighbor : adj[i]) {
                cout << neighbor << " ";
            }
            cout << "\n";
        }
    }
};

int main() {
    int V, E;
    bool isDirected;
    cout << "Enter number of vertices: ";
    cin >> V;
    cout << "Is the graph directed? (1 for yes, 0 for no): ";
    cin >> isDirected;

    Graph g(V, isDirected);

    cout << "Enter number of edges: ";
    cin >> E;
    cout << "Enter edges (u v):\n";
    for (int i = 0; i < E; ++i) {
        int u, v;
        cin >> u >> v;
        g.addEdge(u, v);
    }

    g.display();
    return 0;
}

File: DSA/GRAPH/1_Basics/Representation/AdjacencyListGraph.java
import java.util.*;

public class AdjacencyListGraph {
    private int V;
    private boolean isDirected;
    private List<List<Integer>> adj;

    public AdjacencyListGraph(int V, boolean isDirected) {
        this.V = V;
        this.isDirected = isDirected;
        adj = new ArrayList<>();
        for (int i = 0; i < V; i++)
            adj.add(new ArrayList<>());
    }

    public void addEdge(int u, int v) {
        adj.get(u).add(v);
        if (!isDirected)
            adj.get(v).add(u);
    }

    public void printGraph() {
        System.out.println("Adjacency List:");
        for (int i = 0; i < V; i++) {
            System.out.print(i + ": ");
            for (int neighbor : adj.get(i)) {
                System.out.print(neighbor + " ");
            }
            System.out.println();
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter number of vertices: ");
        int V = scanner.nextInt();
        System.out.print("Is the graph directed? (true/false): ");
        boolean directed = scanner.nextBoolean();

        AdjacencyListGraph graph = new AdjacencyListGraph(V, directed);

        System.out.print("Enter number of edges: ");
        int E = scanner.nextInt();

        System.out.println("Enter edges (u v):");
        for (int i = 0; i < E; i++) {
            int u = scanner.nextInt();
            int v = scanner.nextInt();
            graph.addEdge(u, v);
        }

        graph.printGraph();
        scanner.close();
    }
}
